Added support for multiple API endpoints to `HFClientVLLM`

umarbutler opened the pull request:
This PR adds support for multiple API endpoints being passed to `HFClientVLLM` via the `url` argument, in a backwards-compatible manner, as discussed in [this](https://github.com/stanfordnlp/dspy/issues/554) issue.

arnavsinghvi11 commented on dsp/modules/hf_client.py:
```diff
@@ -116,7 +116,16 @@ def send_hftgi_request_v00(arg, **kwargs):
-        self.url = f"{url}:{port}"
+        
+        if type(url) == list:
+            self.urls = ['/v1'.join(url_.split('/v1')[:-1]) for url_ in url]
```
this seems a bit extra - would it be better to specify some base formatting in the documentation and having a condition check on that instead? feel free to add it!

XenonMolecule replied:
I agree with this. It's unclear why we need to do this splitting and joining.  I feel this might lead to unintuitive behavior, for example, if I don't include /v1/completions in my URLs now it drops the URL completely:

```
>>> urls = ["localhost:8000/v1/completions", "server.com:7410/v1/completions", "mycluster:4000", "oldservice.co:3000/v1/service/v1/completions"]
>>> new_urls = ['/v1'.join(url_.split('/v1')[:-1]) for url_ in urls]
>>> new_urls
['localhost:8000', 'server.com:7410', '', 'oldservice.co:3000/v1/service']
```

This would break the existing way people are inputting URLs if they just convert to list form.

XenonMolecule replied:
Also we might want to add a warning if the user passes a port and list of URLs because the port is now ignored.

arnavsinghvi11 commented:
Thanks @umarbutler ! left a comment, but the rest looks good to me!

okhat commented:
This is pretty cool, I think @XenonMolecule @klopsahlong actually needed this right now, can you check and merge if all looks good @arnavsinghvi11 or @XenonMolecule ?

XenonMolecule commented on dsp/modules/hf_client.py:
```diff
@@ -116,7 +116,16 @@ def send_hftgi_request_v00(arg, **kwargs):
+        
+        elif type(url) == str:
+            self.urls = [f'{url}:{port}']
+        
```
It is not necessary to add in this PR, but a related idea that I think it might be nice to add is the same way that HFClientTGI handles multiple ports.

`self.ports = port if isinstance(port, list) else [port]`

In this case if the user provides 1 URL and a list of ports we can round robin over the ports instead (by constructing self.urls over the URL + all ports)

If they specify both as a list we can just raise an exception and say if you have multiple ports and multiple URLs they need to all be specified in the URL list.

XenonMolecule reviewed (commented):
Left a few comments as well.  Looks good!  The main one that I'd want addressed is the same thing Arnav commented on

umarbutler commented:
@arnavsinghvi11 @XenonMolecule I've removed the unnecessary dropping of `v1/*` from urls ending in `v1/*`. I added it because when I first used `HFClientVLLM`, I passed the full endpoint, which raised an error, yet there was no indication that I shouldn't have done so. It looks like there was a typo in my implementation anyways. It should have been `self.urls = [('/v1'.join(url_.split('/v1')[:-1]) if 'v1/' in url_ else url_) for url_ in url]`.
